src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,657 @@
+                     *       and store the passed in Downstream<? super RR>
+                     *       downstream as an instance field in integrate()
+                     *       and read it in push(R r).
+                     *    2. Allocate a new Gatherer.Downstream<? super R> for
+                     *       each invocation of integrate() which might prove
+                     *       costly.
+                     */
+                    public boolean rightIntegrate(R r, Downstream<? super RR> downstream) {
+                        // The following logic is highly performance sensitive
+                        return (rightGreedy || rightProceed)
+                                && (rightIntegrator.integrate(rightState, r, downstream)
+                                || rightGreedy
+                                || (rightProceed = false));
+                    }
+                }
+
+                return new GathererImpl<T, State, RR>(
+                        State::new,
+                        (leftGreedy && rightGreedy)
+                                ? Integrator.<State, T, RR>ofGreedy(State::integrate)
+                                : Integrator.<State, T, RR>of(State::integrate),
+                        (leftCombiner == Gatherer.defaultCombiner()
+                        || rightCombiner == Gatherer.defaultCombiner())
+                                ? Gatherer.defaultCombiner()
+                                : State::joinLeft,
+                        (leftFinisher == Gatherer.<A, R>defaultFinisher()
+                        && rightFinisher == Gatherer.<AA, RR>defaultFinisher())
+                                ? Gatherer.defaultFinisher()
+                                : State::finish
+                );
+            }
+        }
+    }
+
+    // Public built-in Gatherers and factory methods for them
+
+    /**
+     * Gathers elements into fixed-size windows. The last window may contain
+     * fewer elements than the supplied window size.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2, 3], [4, 5, 6], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowFixed(3)).toList();
+     * }
+     *
+     * @apiNote For efficiency reasons, windows may be allocated contiguously
+     *          and eagerly. This means that choosing large window sizes for
+     *          small streams may use excessive memory for the duration of
+     *          evaluation of this operation.
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into fixed-size windows
+     * @throws IllegalArgumentException when windowSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowFixed(int windowSize) {
```

Comment:
In my opinion, it would be nicer to let `Gatherers` be a shopping window for cool gatherers. Ideally, I think only the docs, methods, parameters, and invariant assertions should be visible here. The rest could be tucked away under the covers.

Reply:
@minborg While I think that would be fantastic from a theoretical standpoint -- however, as we've already seen with windowSliding and windowFixed, the need for performance and correctness will mean that the Gatherer implementations herein will likely not be "short and nice" which would be ideal from a "shopping window" perspective. And remember, it is not always the case that the cheese in the window is real. :)